### PR TITLE
Add stubbing from json mapping file

### PR DESCRIPTION
**Pull request by marcinkp**

### CODE REVIEW

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java [hunk 1]
```diff
@@ -209,7 +254,9 @@ public String duration() {
 
     @Override
     public String toString() {
-      return String.format("Success ratio: %s/%s=%s | Duration: %s", succeed(), total, succeedPercent(), duration());
+      return String
+        .format("Success ratio: %s/%s=%s | Duration: %s", succeed(), total, succeedPercent(),
```
Comment (pun-ky) [inline core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:258]: Uglier

File: wiremock/src/main/java/com/cognifide/aem/stubs/wiremock/WireMockConfig.java [hunk 2]
```diff
@@ -37,21 +37,23 @@
...
+
   private final ResolverAccessor resolverAccessor;
   private final String rootPath;
   private final boolean globalTransformer;
   private final Map<String, Extension> extensions = newLinkedHashMap();
 
 
-  public WireMockConfig(ResolverAccessor resolverAccessor, String rootPath, boolean globalTransformer) {
+  public WireMockConfig(ResolverAccessor resolverAccessor, String rootPath,
+    boolean globalTransformer) {
```
Comment (pun-ky) [inline wiremock/src/main/java/com/cognifide/aem/stubs/wiremock/WireMockConfig.java:48]: I am assuming 160  chars line length

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java [hunk 3]
```diff
@@ -219,8 +266,11 @@ public String toString() {
     @AttributeDefinition(name = "Root Path")
     String resource_paths() default "/var/stubs";
 
-    @AttributeDefinition(name = "Extension")
-    String extension() default ".groovy";
+    @AttributeDefinition(name = "Script Extension")
+    String scriptExtension() default ".groovy";
+
+    @AttributeDefinition(name = "Mappings Extension")
```
Comment (pun-ky) [inline core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:272]: Singular form

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java [hunk 4]
```diff
@@ -108,7 +117,8 @@ public void runAll(Stubs<?> runnable) {
     LOG.info("Running AEM Stubs scripts result: {}", result);
   }
 
-  private void runAllEach(Resource resource, RunAllResult result, Stubs<?> runnable, ResourceResolver resolver) {
+  private void runAllEach(Resource resource, RunAllResult result, Stubs<?> runnable,
+    ResourceResolver resolver) {
```
Comment (pun-ky) [inline core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:121]: Why breaking line?

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java [hunk 5]
```diff
@@ -121,10 +131,20 @@ private void runAllEach(Resource resource, RunAllResult result, Stubs<?> runnabl
   @Override
   public Optional<Stubs<?>> findRunnable(String path) {
     return runnables.stream()
-      .filter(runnable -> wildcardMatch(path, format("%s/%s/**/*%s", getRootPath(), runnable.getId(), getExtension())))
+      .filter(runnable -> isStubScript(path, runnable) || isMappingFile(path, runnable))
       .findFirst();
   }
 
+  private boolean isStubScript(String path, Stubs<?> runnable) {
```
Comment (pun-ky) [inline core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:138]: DRY rule

Fully qualified : StubScriptManager.isStubScript.. You are repeating Stub... That's common java devs practice which I dislike very much. Context matters. Why not just isScript..

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java [hunk 6]
```diff
@@ -145,29 +169,50 @@ public String getRootPath() {
...
   public void onChange(List<ResourceChange> changes) {
-    final List<String> scriptPaths = changes.stream()
-      .map(ResourceChange::getPath)
-      .filter(this::isRunnable)
-      .collect(Collectors.toList());
+    final List<String> mappingPaths = getMappingPaths(changes);
+
+    if (!mappingPaths.isEmpty()) {
+      resetAll(mappingPaths);
+      return;
```
Comment (pun-ky) [inline core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:181]: Why distinguishing paths which in the end are reseting all...

If reloading single mapping file is not possible maybe we should remove on change modes at all

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java [hunk 7]
```diff
@@ -121,10 +131,20 @@ private void runAllEach(Resource resource, RunAllResult result, Stubs<?> runnabl
...
       .findFirst();
   }
 
+  private boolean isStubScript(String path, Stubs<?> runnable) {
+    return wildcardMatch(path,
+      format("%s/%s/**/*%s", getRootPath(), runnable.getId(), getExtension()));
+  }
+
+  private boolean isMappingFile(String path, Stubs<?> runnable) {
+    return wildcardMatch(path,
```
Comment (pun-ky) [inline core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:144]: isMapping

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java (hunk 3 above)
Comment (pun-ky) [reply core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:272]: The extensions are configurable however the subfolders 'mappings' and 'scripts' are hardcoded

File: wiremock/src/main/java/com/cognifide/aem/stubs/wiremock/WireMockFileSource.java [hunk 8]
```diff
@@ -87,9 +91,27 @@ public URI getUri() {
...
 
+  private List<TextFile> listFiles(String folderPath) {
+    return resolverAccessor.resolve(resourceResolver -> {
+      return StreamUtils.from(resourceResolver.getResource(folderPath).getChildren().iterator())
+        .flatMap(resource -> {
+          if (resource.isResourceType("sling:Folder")) {
+            return listFiles(resource.getPath()).stream();
+          } else {
+            if (resource.isResourceType("nt:file")) {
+              return Stream.of(new WireMockFileSource(this.resolverAccessor, folderPath).getTextFileNamed(resource.getName()));
```
Comment (pun-ky) [inline wiremock/src/main/java/com/cognifide/aem/stubs/wiremock/WireMockFileSource.java:105]: There is our constant for that node type

File: wiremock/src/main/content/jcr_root/var/stubs/wiremock/mappings/samples/hello-world.json [hunk 9]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (pun-ky) [inline wiremock/src/main/content/jcr_root/var/stubs/wiremock/mappings/samples/hello-world.json:1]: Nice one! 🙂

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java (hunk 7 above)
Comment (marcinkp) [reply core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:139]: done

File: wiremock/src/main/content/jcr_root/var/stubs/wiremock/mappings/samples/hello-world.json (hunk 9 above)
Comment (marcinkp) [reply wiremock/src/main/content/jcr_root/var/stubs/wiremock/mappings/samples/hello-world.json]: thanks :)

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java (hunk 4 above)
Comment (marcinkp) [reply core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java]: done

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java (hunk 3 above)
Comment (marcinkp) [reply core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:242]: done

**Comment by marcinkp** [comment]
I implemented change you suggested. No subfolder

File: core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java (hunk 6 above)
Comment (marcinkp) [reply core/src/main/java/com/cognifide/aem/stubs/core/script/ConfigurableStubScriptManager.java:181]: hmm. for mappings for sure it is needed - restart all

**Pull request merged**